kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -425,240 +289,374 @@ private void writeApiVersionsResponse(ChannelHandlerContext ctx, DecodedRequestF
+     */
     @Override
     public String clientHost() {
-        if (haProxyMessage != null) {
-            return haProxyMessage.sourceAddress();
+        final SelectingServer selectingServer = proxyChannelStateMachine
+                .enforceInSelectingServer(NET_FILTER_INVOKED_IN_WRONG_STATE + proxyChannelStateMachine.currentState());
+        if (selectingServer.haProxyMessage() != null) {
+            return selectingServer.haProxyMessage().sourceAddress();
         }
         else {
-            SocketAddress socketAddress = inboundCtx.channel().remoteAddress();
-            if (socketAddress instanceof InetSocketAddress) {
-                return ((InetSocketAddress) socketAddress).getAddress().getHostAddress();
+            SocketAddress socketAddress = clientCtx().channel().remoteAddress();
+            if (socketAddress instanceof InetSocketAddress inetSocketAddress) {
+                return inetSocketAddress.getAddress().getHostAddress();
             }
             else {
                 return String.valueOf(socketAddress);
             }
         }
     }
 
+    /**
+     * Accessor exposing the client port to a {@link NetFilter}.
+     * <p>Called by the {@link #netFilter}.</p>
+     * @return The client port.
+     * @throws IllegalStateException if {@link #proxyChannelStateMachine} is not {@link SelectingServer}.
+     */
     @Override
     public int clientPort() {
-        if (haProxyMessage != null) {
-            return haProxyMessage.sourcePort();
+        final SelectingServer selectingServer = proxyChannelStateMachine
+                .enforceInSelectingServer(NET_FILTER_INVOKED_IN_WRONG_STATE + proxyChannelStateMachine.currentState());
+        if (selectingServer.haProxyMessage() != null) {
+            return selectingServer.haProxyMessage().sourcePort();
         }
         else {
-            SocketAddress socketAddress = inboundCtx.channel().remoteAddress();
-            if (socketAddress instanceof InetSocketAddress) {
-                return ((InetSocketAddress) socketAddress).getPort();
+            SocketAddress socketAddress = clientCtx().channel().remoteAddress();
+            if (socketAddress instanceof InetSocketAddress inetSocketAddress) {
+                return inetSocketAddress.getPort();
             }
             else {
                 return -1;
             }
         }
     }
 
+    /**
+     * Accessor exposing the source address to a {@link NetFilter}.
+     * <p>Called by the {@link #netFilter}.</p>
+     * @return The source address.
+     * @throws IllegalStateException if {@link #proxyChannelStateMachine} is not {@link SelectingServer}.
+     */
     @Override
     public SocketAddress srcAddress() {
-        return inboundCtx.channel().remoteAddress();
+        proxyChannelStateMachine.enforceInSelectingServer(NET_FILTER_INVOKED_IN_WRONG_STATE + proxyChannelStateMachine.currentState());
```

Comment:
Seems a bit odd to have to pass the current state in the message like this. Why isn't this just an `proxyChannelStateMachine.enforceInSelectingServer()`, or at most `proxyChannelStateMachine.enforceInSelectingServer(NET_FILTER_INVOKED_IN_WRONG_STATE)` and have that method add the current state to the message?

Reply:
Yeah, it felt odd to encode the details about the `NetFilter` usage in the state machine, which is why I didn't just inline the message. 

I wasn't sure if having the state machine arbitrarily append the current state made much sense as an API either. That said I am happy to for it do so, I think that makes the calls sites a lot cleaner.

Reply:
https://github.com/kroxylicious/kroxylicious/pull/1549/files#r1828434340 was where I was wondering previously

Reply:
resolved in fa3feae5c361c0a316d30a263234cff78f8ef475